diff -ur o-ruby-fcgi-0.8.5/ext/fcgi/fcgi.c ruby-fcgi-0.8.5/ext/fcgi/fcgi.c
--- o-ruby-fcgi-0.8.5/ext/fcgi/fcgi.c	2005-03-03 14:21:55.980081000 -0700
+++ ruby-fcgi-0.8.5/ext/fcgi/fcgi.c	2005-03-03 15:25:42.814314608 -0700
@@ -157,9 +157,11 @@
 
 #define CHECK_STREAM_ERROR(stream) {\
   int err = FCGX_GetError(stream);\
+	extern int errno; \
   if (err) {\
     if (err > 0) {\
-      rb_sys_fail(0);\
+			errno = err;\
+      rb_sys_fail(NULL);\
     }\
     else {\
       switch (err) {\
