diff -ur ruby-fcgi-0.8.6-o/ext/fcgi/fcgi.c ruby-fcgi-0.8.6/ext/fcgi/fcgi.c
--- ruby-fcgi-0.8.6-o/ext/fcgi/fcgi.c	2005-03-31 18:21:41.000000000 -0700
+++ ruby-fcgi-0.8.6/ext/fcgi/fcgi.c	2005-12-28 16:26:22.000000000 -0700
@@ -8,6 +8,7 @@
 #include <sys/types.h>
 #include <unistd.h>
 #include <errno.h>
+#include <fcntl.h>
 
 #include "ruby.h"
 #ifdef HAVE_FASTCGI_FCGIAPP_H
@@ -74,6 +75,15 @@
     char **env;
     VALUE obj,key, value;
     char *pkey,*pvalue;
+    int flags, fd;
+
+    /* Unset NONBLOCKING */
+    fd = ((FCGX_Request*) req)->ipcFd;
+    flags = fcntl(fd, F_GETFL);
+
+    if (flags & O_NONBLOCK) {
+       fcntl(fd, F_SETFL, flags & ~O_NONBLOCK);
+    }
     
     obj = Data_Make_Struct(self, fcgi_data, fcgi_mark, fcgi_free_req, data);
     data->req = req;
Only in ruby-fcgi-0.8.6/ext/fcgi: fcgi.c.orig
Only in ruby-fcgi-0.8.6/ext/fcgi: fcgi.c.rej
Only in ruby-fcgi-0.8.6/ext/fcgi: fcgi.c~
